docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,214 @@
+
+#### Installation
+
+ASO is a self-contained component published independently of CAPZ. When CAPZ is relying on ASO for critical functionality, CAPZ's installation process should also automatically install ASO.
+
+When installing CAPZ with `clusterctl init`, the Kubernetes resources defined by CAPZ's `infrastructure-components.yaml` published on the corresponding release are installed. ASO can also be [installed with a YAML manifest](https://azure.github.io/azure-service-operator/introduction/installing-from-yaml/) published for each of its releases describing its components that can be incorporated with CAPZ's `infrastructure-components.yaml`. This method installs ASO in the `azureserviceoperator-system` namespace and expects a Secret named `aso-controller-settings` to exist with Service Principal credentials. The easiest and most flexible way to customize that manifest further would likely be to generate a manifest from ASO's Helm chart with any necessary values overridden.
+
+Other than integrating with CAPZ's `infrastructure-components.yaml`, the `capz-controller-manager` container itself could assume responsibility for installing ASO by using the Helm Go library to install the ASO chart when it starts. Installing ASO this way allows CAPZ to hook in additional logic, such as to check if ASO is already installed in the cluster or to verify ASO is healthy after it's installed, which would not be possible if a static manifest was embedded in `infrastructure-components.yaml`.
+
+In terms of specific options, ASO offers both single- and multi-operator configurations. The single-operator option is the default, and assumes a single ASO installation managing resources for the whole cluster. The multi-operator method allows several ASO installations to coexist, each monitoring resources in a unique set of namespaces to enable multitenancy. ASO's single-operator configuration also allows multitenancy with more tightly-scoped credentials. Overall, a single ASO installation will likely be the easier path forward since it's consistent with CAPZ's multitenancy model and doesn't involve managing ASO's CRDs and webhooks across multiple installations.
+
+#### Caveats
+
+CAPZ interacts with some Azure services that do not represent infrastructure, and thus cannot be represented in ASO. Resource Health, for example, is "reconciled" by CAPZ currently by getting a resource's health status and reflecting that in the corresponding CAPZ resource, but does not create or update any distinct Azure resources. The new SDK could be used to implement this existing functionality without affecting other service interfaces' use of ASO.
+
+Also, use of the `clusterctl move` command will require extra manual steps to move ASO resources as documented here: https://azure.github.io/azure-service-operator/introduction/frequently-asked-questions/#what-is-the-best-practice-for-transferring-aso-resources-from-one-cluster-to-another.
```

Comment:
let's expand on this part, this is a big one

Reply:
From the PR where AAD pod identity was added, I see the same question about `clusterctl move` asked here: https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/977#issuecomment-730662055

That question seems to be solved with https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/977#issuecomment-740188925 but I'm not exactly sure what that means. What is the "custom type" in this context?

Reply:
Because CAPI has automated moving CAPI resources from one management cluster to another, we should be able to extend that to ASO too.

For AAD pod identity it was more tricky because of the one-to-many aspect (one identity can be shared by multiple clusters, so moving a cluster doesn't mean you should move its identity with it, a cluster doesn't "own" its identity). For ASO it should be a bit easier because there is a strong Ownership chain (unless maybe you share a VNet object or ResourceGroup object between two clusters, that edge case needs to be handled separately). 

So in theory when we move the CAPI cluster we should also move ASO CRDs to the target cluster. The one thing we may need to handle is "pausing" reconciliation of the ASO resources during the move by annotating them with `serviceoperator.azure.com/reconcile-policy: skip` (this could be part of the CAPZ controller whenever the CAPZ resources are paused). Then, we also need to make sure the ASO resources get moved following the following rules, either via `OwnerReference` or labeling all of them with `clusterctl.cluster.x-k8s.io/move-hierarchy`:

https://cluster-api.sigs.k8s.io/clusterctl/provider-contract.html#move

Overall, we should make this as automated as possible so the user doesn't need to know/worry about ASO since from the user's standpoint it's an implementation details and not something they should need to care about.

Reply:
Ah, I wasn't sure ASO resources would be easy to hook into move. Based on my read of that link, I thought CAPI would only move CAPI objects, Secrets, and ConfigMaps. But

> All the Kind defined in one of the CRDs installed by clusterctl using `clusterctl init` (identified via the `clusterctl.cluster.x-k8s.io` label)

sounds like that would also include the ASO CRDs if they're defined in infrastructure-components.yaml. And maybe we could apply that label ourselves if we decide to install ASO from inside the CAPZ controller manager?

I'll dig into this more and circle back.

Reply:
It looks like as long as an owner ref is set on the ASO resource group and other top-level resources, then `clusterctl` will detect and move all of the resources automatically. The steps to change the reconcile-policy annotation to "skip" and back to "manage" still must be done manually though before and after the `clusterctl move`. We'll still need to program them for our e2e tests at least. 

Things seem to work okay even if I let the annotation stay the same since `clusterctl` seems to be quick about creating then deleting the objects, but it seems prudent to communicate those steps as being necessary anyway to make sure two ASOs are never fighting over the same resource.

I've updated the doc here to be a bit more verbose.

Reply:
> The steps to change the reconcile-policy annotation to "skip" and back to "manage" still must be done manually though before and after the clusterctl move. We'll still need to program them for our e2e tests at least.

Do you have any ideas on how we could automate this? Perhaps in the CAPZ controller?

Reply:
> Things seem to work okay even if I let the annotation stay the same since clusterctl seems to be quick about creating then deleting the objects, but it seems prudent to communicate those steps as being necessary anyway to make sure two ASOs are never fighting over the same resource.

Yeah the risk here is twofold:
1. If the user deletes the cluster on the old cluster post move, if you don't have the `skip` or `detach-on-delete` annotations, the deletion of the ASO resources will trigger deletion of the corresponding ARM resources.
2. If the user modifies the cluster (and thus the ASO resources) on the new cluster post move, it's possible that the old and new ASO resources fight with one another as they have different goal states.

number 2 is probably low-risk but number 1 is scary.

Reply:
> Do you have any ideas on how we could automate this? Perhaps in the CAPZ controller?

I've thought about this a bit and I think the missing (to me) pieces are:
1. How would CAPZ detect a move is _about to_ happen?
2. How would CAPZ then block the move to do its ASO annotation swizzling, then unblock the move?

Reply:
> 1. How would CAPZ detect a move is _about to_ happen?

Actually, this seems like a special case of "What happens when a cluster is (un)paused?" which is a question we have yet to address. In that case, CAPZ should probably set `reconcile-policy: skip` on all the ASO resources which is the same as what we're looking to accomplish here.

This would involve invoking a reconciliation in CAPZ when a Cluster is newly paused which I don't think happens today, but I think if we have a completely separate path for that like `reconcileNormal` vs. `reconcileDelete` then it seems like we can ensure no "actual" reconciliation happens then.

> 2. How would CAPZ then block the move to do its ASO annotation swizzling, then unblock the move?

I was looking to see if adding a new finalizer to the AzureCluster would block it from being deleted during the move to allow CAPZ to update the annotation on the ASO resources, but `clusterctl move` plows right through all the finalizers. https://github.com/kubernetes-sigs/cluster-api/blob/v1.3.5/cmd/clusterctl/client/cluster/mover.go#L1136-L1141. @CecileRobertMichon Does that seem like something we could change in CAPI, to have `clusterctl move` either only remove a specific set of finalizers or create some sort of naming convention for finalizers that `clusterctl move` should not automatically delete?

Reply:
Hm, editing the annotations during delete won't work for `clusterctl move` since the delete happens after create in the new cluster.

Reply:
How does CAPZ prevent the backing Azure resources being deleted today when move happens? Presumably `kubectl delete` or equivalent is called on the CAPZ resources post-move. Why doesn't that trigger delete of the resources in Azure?

Reply:
I think the `spec.paused: true` that `clusterctl` sets at the beginning of a move does that. If a cluster is paused, I don't think any of the CAPZ controllers will be invoked for resources part of that cluster, even for delete.

Reply:
Gotcha, the trick/issue is that `clusterctl` sets that generically and so can't set the Azure specific `skip` annotation on ASO's resources via `clusterctl` since it has no idea about them. Thanks for explaining.

Reply:
@CecileRobertMichon Do you think it would be too disruptive to eliminate the assumption in CAPI that pausing a cluster is instantaneous? Perhaps by adding indicating that a cluster is paused in its status if some `wait-for-pause` annotation exists on the Cluster? So `clusterctl move` would set `spec.paused` then wait for `status.paused` to be set before continuing.

Reply:
Yep, opened: https://github.com/kubernetes-sigs/cluster-api/issues/8473